internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,185 @@
+	globalRateLimit time.Duration
+}
+
+// New returns a new RateLimiter
+func New() *RateLimiter {
+
+	return &RateLimiter{
+		buckets: make(map[string]*Bucket),
+	}
+}
+
+// getBucket retrieves or creates a bucket
+// ratelimiter is expected to be locked when calling this
+func (r *RateLimiter) getBucket(key string) *Bucket {
+
+	if bucket, ok := r.buckets[key]; ok {
+		return bucket
+	}
+
+	b := &Bucket{remaining: 1, r: r, Key: key}
+	r.buckets[key] = b
+	return b
+}
+
+// LockBucket Locks until a request can be made
+func (r *RateLimiter) LockBucket(path string) *Bucket {
+
+	bucketKey := ParseURL(path)
+
+	r.Lock()
+	b := r.getBucket(bucketKey)
+	r.Unlock()
+
+	b.mu.Lock()
+
+	// If we ran out of calls and the reset time is still ahead of us
+	// then we need to take it easy and relax a little
+	for b.remaining < 1 && b.reset.After(time.Now()) {
+		time.Sleep(b.reset.Sub(time.Now()))
+	}
+
+	// Lock and unlock to check for global ratelimites after sleeping
+	r.Lock()
+	r.Unlock()
+
+	b.remaining--
+	return b
+}
+
+// Bucket represents a ratelimit bucket, each bucket gets ratelimited individually (-global ratelimits)
+type Bucket struct {
+	Key string
+
+	mu        sync.Mutex
+	remaining int
+	limit     int
+	reset     time.Time
+	r         *RateLimiter
+}
+
+// Release unlocks the bucket and reads the headers to update the buckets ratelimit info
+// and locks up the whole thing in case if there's a global ratelimit.
+func (b *Bucket) Release(headers http.Header) error {
+
+	defer b.mu.Unlock()
+	if headers == nil {
+		return nil
+	}
+
+	remaining := headers.Get("X-RateLimit-Remaining")
+	reset := headers.Get("X-RateLimit-Reset")
+	global := headers.Get("X-RateLimit-Global")
+	retryAfter := headers.Get("Retry-After")
+
+	// If it's global just keep the main ratelimit mutex locked
+	if global != "" {
+		parsedAfter, err := strconv.Atoi(retryAfter)
+		if err != nil {
+			return err
+		}
+
+		// Lock it in a new goroutine so that this isn't a blocking call
+		go func() {
+			// Make sure if several requests were waiting we don't sleep for n * retry-after
+			// where n is the amount of requests that were going on
+			sleepTo := time.Now().Add(time.Duration(parsedAfter) * time.Millisecond)
+
+			b.r.Lock()
+
+			sleepDuration := sleepTo.Sub(time.Now())
+			if sleepDuration > 0 {
+				time.Sleep(sleepDuration)
+			}
+
+			b.r.Unlock()
+		}()
+
+		return nil
+	}
+
+	// Update reset time if either retry after or reset headers are present
+	// Prefer retryafter because it's more accurate with time sync and whatnot
+	if retryAfter != "" {
+		parsedAfter, err := strconv.ParseInt(retryAfter, 10, 64)
+		if err != nil {
+			return err
+		}
+		b.reset = time.Now().Add(time.Duration(parsedAfter) * time.Millisecond)
+
+	} else if reset != "" {
+		unix, err := strconv.ParseInt(reset, 10, 64)
+		if err != nil {
+			return err
+		}
+
+		// Add a second to account for time desync and such
+		b.reset = time.Unix(unix, 0).Add(time.Second)
```

Comment:
Also, can you play with this value.. You mentioned a few milliseconds don't work, what about 100*time.Millisecond, 200? 500?